components/sup/src/manager/service/config.rs

```diff
@@ -60,30 +63,71 @@ pub struct ServiceConfig {
     // Set to 'true' if we have data that needs to be sent to a configuration file
     #[serde(skip_deserializing)]
     pub needs_write: bool,
+    #[serde(skip_deserializing)]
+    supported_bindings: Vec<(String, String)>,
 }
 
 impl ServiceConfig {
     /// Takes a new package and a new census list, and returns a ServiceConfig. This function can
     /// fail, and indeed, we want it to - it causes the program to crash if we can not render the
     /// first pass of the configuration file.
-    pub fn new(service_group: &str,
-               package: &Package,
-               cl: &CensusList,
-               bindings: &[String])
-               -> Result<ServiceConfig> {
-        let cfg = try!(Cfg::new(package));
-        let bind = try!(Bind::new(bindings, &cl));
+    pub fn new<T: AsRef<Path>>(package: &Package,
+                               runtime: &RuntimeConfig,
+                               config_root: T,
+                               bindings: &[String])
+                               -> Result<ServiceConfig> {
         Ok(ServiceConfig {
-            pkg: Pkg::new(&package.pkg_install),
+            pkg: Pkg::new(package, runtime)?,
             hab: Hab::new(),
             sys: Sys::new(),
-            cfg: cfg,
-            svc: Svc::new(service_group, cl),
-            bind: bind,
+            cfg: Cfg::new(package, config_root)?,
+            svc: Svc::default(),
+            bind: Bind::default(),
             needs_write: true,
+            supported_bindings: Self::split_bindings(bindings)?,
         })
     }
 
+    fn split_bindings(bindings: &[String]) -> Result<Vec<(String, String)>> {
```

Comment:
The earlier that we validate input the better. I would move this function to that area in a future refactor if I were you for sure!